Assets/Scripts/InputManager/InputComponent.cs

```diff
@@ -9,73 +13,94 @@ public class InputComponent
         public float verticalVal { get; private set; }
         private bool shoot { get; set; }
 
-        public PlayerController playerController { private get; set; }
+        List<InputAction> actions;
+        public PlayerController playerController { get; set; }
         public InputComponentScriptable inputComponentScriptable { private get; set; }
 
+        //public InputComponent()
+        //{
+        //    actions.Add(new SpawnAction(playerController.getSpawnPos()));
+        //}
+
         public void OnUpdate()
         {
-            if (Input.GetKeyDown(inputComponentScriptable.fireKey))
-            {
-                playerController.setFireState();
-            }
-            else if (Input.GetKeyUp(inputComponentScriptable.fireKey))
+            if (GameManager.Instance.currentState.gameStateType == GameStateType.Pause) return;
+
+            actions = new List<InputAction>();
+
+            if(Input.GetKeyDown(inputComponentScriptable.fireKey))
             {
-                playerController.SetStateFales(playerController.characterFireState);
+                actions.Add(new FireAction());
             }
 
             MoveForward();
             MoveBackward();
             TurnLeft();
             TurnRight();
 
-            if (horizontalVal != 0 || verticalVal != 0)
-            {
-                playerController.setMoveState();
-            }
-            else if (horizontalVal == 0 && verticalVal == 0)
+            if (horizontalVal == 0 && verticalVal == 0)
             {
-                playerController.setIdleState();
+                if (playerController.currentState != playerController.characterIdleState)
+                    actions.Add(new IdleAction());
             }
+
+            ReplayManager.Instance.SaveCurrentQueueData(actions);
```

Comment:
ideally the replay manager should read the data from input manager and not individual controller